### DISCUSSION TITLE
Different CSS inject order in development and production (MUI/emotion and Tailwind)

### DISCUSSION THREAD

**Discussion by cibulka** [vercel/next.js#32565] [Help]
Hi! I'm trying to make [TailwindCSS](https://tailwindcss.com/) play nicely with [MUI](https://mui.com/). 

The problem I'm facing is a different behaviour between development and production. Take this:

```
import Skeleton from '@mui/material/Skeleton'
<Skeleton className="w-20 h-20" variant="circular" />
```

On **development**, Tailwind gets higher priority (is injected later, I assume), so it overwrites hardcoded MUI styles. That is the way I would prefer (especially as it seems that [one can not remove CSS rules in current version of MUI](https://stackoverflow.com/questions/64818441/remove-css-rule-for-material-ui-component-in-theme-overrides)).

<img width="991" alt="Screenshot 2021-12-16 at 12 00 56" src="https://user-images.githubusercontent.com/3989833/146359943-448e13fc-1845-4d95-af63-4b3eb724995c.png">

On **production**, MUI (resp. emotion) takes over, which slaps `height: 1.2rem` over the Skeleton and squishes it.

<img width="1132" alt="Screenshot 2021-12-16 at 12 01 46" src="https://user-images.githubusercontent.com/3989833/146360034-64e1543c-f6aa-4198-936c-a165fdbf9753.png">

So yeah, not the best dev experience. :) Just imagine the number of bugs you find out *only after you deploy the thing*. 

[I've created the repo where you can reproduce the problem](https://github.com/cibulka/reproduction-next-tailwind-mui). You know the drill:

```
npm install
npm run dev # development
npm run build && npm run start # production
```

Is this expected behaviour? Can I regulate this somehow (e.g. give Tailwind bigger priority)? 

Also I assume that out of the three, I should ask those questions to Next.js, as it does the CSS importing.

Thank you!

**Comment by cibulka** [comment]
MUI actually exposes `StyledEngineProvider` with a parameters `injectFirst` that allows exactly that (so the MUI styles have lower priority). That fixes my problem! 

```
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles';

function MyApp({ Component, pageProps }: AppProps) {

  return (
    <StyledEngineProvider injectFirst>
      <ThemeProvider theme={muiTheme}>
        <Component {...pageProps} />
      </ThemeProvider>
    </StyledEngineProvider>
  );
```

The full solution is available in the repo above, branch `withInjectOrder`. 

I still think thought that the inconsistency between development and production is an issue, so I'm leaving this thread open.

**Reply by Ahmed-Nasser-99** [reply]
Thanks so much .
I have just encountered the same problem and this thread saved hours of searching.

**Reply by iwanhae** [reply]
You saved whole my day! Thx!

**Reply by david-benes** [reply]
You saved my sanity! :D Thank you for this.

**Reply by danbockapps** [reply]
Still works in August 2023

**Reply by JackCotter** [reply]
Still working now. Thank you, saved me so much time.

**Reply by cedricmarcelino** [reply]
Still works 2024

**Reply by hnykda** [reply]
Just awesome, thanks. Works on Nextjs 14 and tailwind 3.4

**Reply by avega93** [reply]
Still works April 2024

**Reply by yizzycool** [reply]
Thank you so much!
This solution also works with **Next.js 14** + **CSS Modules**.

**Reply by abrzezina** [reply]
🥇
